src/openloco/localisation/stringmgr.cpp

```diff
@@ -1,23 +1,563 @@
+    const uint16_t USER_STRINGS_START = 0x8000;
+    const uint16_t USER_STRINGS_END   = USER_STRINGS_START + NUM_USER_STRINGS;
+
+    const uint16_t NUM_TOWN_NAMES     = 345;
+    const uint16_t TOWN_NAMES_START   = 0x9EE7;
+    const uint16_t TOWN_NAMES_END     = TOWN_NAMES_START + NUM_TOWN_NAMES;
+
     static loco_global<char * [0xFFFF], 0x005183FC> _strings;
+    static loco_global<char [NUM_USER_STRINGS][USER_STRING_SIZE], 0x0095885C> _userStrings;
 
     const char* get_string(string_id id)
     {
-        return _strings[id];
+        printf("Fetching string %d\n", id);
+        char* str = _strings[id];
+        printf("Found at %p\n", str);
+
+        return str;
+    }
+
+    // TODO: decltype(value)
+    static char* format_comma(uint32_t value, char* buffer)
+    {
+        registers regs;
+        regs.eax = value;
+        regs.edi = (uint32_t) buffer;
+
+        call(0x00495F35, regs);
+        return (char*) regs.edi;
+    }
+
+    // TODO: decltype(value)
+    static char* format_int(uint32_t value, char* buffer)
+    {
+        registers regs;
+        regs.eax = value;
+        regs.edi = (uint32_t) buffer;
+
+        call(0x495E2A, regs);
+        return (char*) regs.edi;
+    }
+
+    // TODO: decltype(value)
+    static char* formatNumeric_4(uint16_t value, char* buffer)
+    {
+        registers regs;
+        regs.eax = (uint32_t) value;
+        regs.edi = (uint32_t) buffer;
+
+        call(0x4963FC, regs);
+        return (char*) regs.edi;
+    }
+
+    // TODO: decltype(value)
+    static char* format_comma2dp32(uint32_t value, char* buffer)
+    {
+        registers regs;
+        regs.eax = value;
+        regs.edi = (uint32_t) buffer;
+
+        call(0x4962F1, regs);
+        return (char*) regs.edi;
+    }
+
+    static char* formatDayMonthYearFull(uint32_t value, char* buffer)
+    {
+        registers regs;
+        regs.eax = (uint32_t) value;
+        regs.edi = (uint32_t) buffer;
+
+        call(0x495CB5, regs);
+        return (char*) regs.edi;
+    }
+
+    static char* formatMonthYearFull(uint32_t value, char* buffer)
+    {
+        registers regs;
+        regs.eax = (uint32_t) value;
+        regs.edi = (uint32_t) buffer;
+
+        call(0x495D77, regs);
+        return (char*) regs.edi;
     }
 
-    void format_string(char* buffer, string_id id, void* args)
+    static char* formatMonthYearAbbrev_0(uint32_t value, char* buffer)
     {
         registers regs;
-        regs.eax = id;
-        regs.edi = (uint32_t)buffer;
-        regs.ecx = (uint32_t)args;
-        call(0x004958C6, regs);
+        regs.eax = (uint32_t) value;
+        regs.edi = (uint32_t) buffer;
+
+        call(0x495DC7, regs);
+        return (char*) regs.edi;
+    }
+
+    static char* format_string(char* buffer, string_id id, argswrapper &args);
+
+    static char* format_string_part(char* buffer ,const char* sourceStr, argswrapper &args)
+    {
+        while (uint8_t ch = *sourceStr++)
```

Comment:
```c++
        while (true)
        {
            uint8_t ch = *sourceStr++;
```

This (seemingly?) stopped the assert from getting hit for me

Reply:
Thanks for this. I'd been debugging this for a long time, and it turns out it was a C-ism doing my in all along. :smile: